oak_sev_guest/src/ghcb.rs

```diff
@@ -614,3 +667,73 @@ fn reset_slice(slice: &mut [u8]) {
         *byte = 0;
     }
 }
+
+#[repr(C)]
+#[derive(AsBytes, FromBytes, FromZeroes)]
+struct PageStateChangeHeader {
+    cur_entry: u16,
+    end_entry: u16,
+    _reserved: u32,
+}
+
+/// Page State Change structure.
+///
+/// See section 4.1.6 in <https://www.amd.com/system/files/TechDocs/56421-guest-hypervisor-communication-block-standardization.pdf>.
+#[repr(C)]
+#[derive(AsBytes, FromBytes, FromZeroes)]
+struct PageStateChange {
+    header: PageStateChangeHeader,
+    entry: [u64; 253],
+}
+static_assertions::assert_eq_size!(PageStateChange, [u8; SHARED_BUFFER_SIZE]);
+
+/// Page State Change Entry.
+///
+/// See Table 9 in <https://www.amd.com/system/files/TechDocs/56421-guest-hypervisor-communication-block-standardization.pdf>.
+struct PageStateChangeEntry<S: NotGiantPageSize> {
+    page_operation: PageAssignment,
+
+    /// Physical frame in guest memory for the operation.
+    gfn: PhysFrame<S>,
+
+    /// Input: offset, in 4K increments, on which to bein the page state change
+    /// operation. Output: offset of the current page in 4K increments that
+    /// has been successfully processed.
+    current_page: u16,
+}
+
+impl<S: NotGiantPageSize> From<PageStateChangeEntry<S>> for u64 {
+    fn from(value: PageStateChangeEntry<S>) -> Self {
+        // [63:57]: Reserved, must be zero. Reset the whole thing to zero.
+        let mut entry = 0u64;
+        // [56]: Page size.
+        entry |= match S::SIZE {
+            Size4KiB::SIZE => (SevPageSize::Page4KiB as u64) << 56,
+            Size2MiB::SIZE => (SevPageSize::Page2MiB as u64) << 56,
+            _ => unreachable!("Unexpected non-giant page size (not 4 KiB or 2 MiB)"),
+        };
+        // [55:52]: Page operation.
+        entry |= (value.page_operation as u64) << 52;
+        // [51:12]: Guest physical frame number
+        // PhysFrame guarantees that the address is properly aligned.
+        entry |= value.gfn.start_address().as_u64();
+        // [11:0]: Current page.
+        entry |= (value.current_page as u64) & 0x800;
```

Comment:
If this is trying to assign the 12 least significant bits should it be `& 0xFFF`?